mk/gx.mk

```diff
@@ -2,8 +2,7 @@ gx-path = gx/ipfs/$(shell gx deps find $(1))/$(1)
 
 # Rebuild the lockfile iff it exists.
 gx-deps: $(wildcard gx-lock.json)
-	gx install --global
-	if test -e gx-lock.json; then gx-go rw --fix && gx lock-install; else rm -rf vendor && gx-go rw; fi
+	if test -e gx-lock.json; then gx-go rw --undo && gx lock-install; else rm -rf vendor && gx install --global && gx-go rw; fi
```

Comment:
> I would move the dependency installation to the lock target. Reinstalling the exact same dependency here is redundant.

The lock target just generates the lockfile, nothing more. `gx-deps` is the "install the gx-deps" target so it's definitely the right place to put this.

>  Also, I may be mistaken, but I think it will cause problems if we do as you suggest and (a) remove the vender directory or (b) replace the symlinks in the vendor directory with symlinks to packages in your go path.

Replacing the symlink definitely works.

Removing the entire vendor directory doesn't *quite* work (intentionally) due to the `github.com/gxed/go-require-gx` import in `cmd/ipfs/ipfs.go`. However, if you remove *that* it should work as long as none of our downstream packages have made breaking changes.

> Also removing the vender directory without promoting the user seams dangerous (there might be stuff in there not related to gx) so it seams that it would be a better idea to just abort if the vender directory exists and let the user deal with it.

In general, vendor directories are managed by whatever dependency management system is in-place, not manually by the user. I don't expect users will be surprised.

I'm worried that *not* removing it will lead to even more confusion. If the user does a `gx-go uw`, they'll end up using the "locked" deps instead of the normal ones.

Reply:
@Stebalien 

> The lock target just generates the lockfile, nothing more.

My concern is that `gx-deps` target is run automatically by the `build` and `install` target.  I never linked this and now it may do undesirable things.  I would be happier if there was a way to tell the build system _not_ to rerun the `gx-deps` target on `make build` or `make install` for the cases when you are doing something slightly unusual and know what you are doing.  Otherwise if this becomes an issue I may end up having to hack the Makefile.

> Replacing the symlink definitely works.

So if you replace the symbolic link, running the gx-deps target won't overright the content that is already in there?

> In general, vendor directories are managed by whatever dependency management system is in-place, not manually by the user. I don't expect users will be surprised.

Okay, I withdraw my objection.

Reply:
> So if you replace the symbolic link, running the gx-deps target won't overright the content that is already in there?

It *should* have worked but apparently we need to fix a bug in gx first.

Reply:
https://github.com/whyrusleeping/gx/issues/215